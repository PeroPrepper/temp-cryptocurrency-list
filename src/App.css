.App {
  text-align: center;
}

.Menu-icon {
  margin-left: 10px;
  margin-right: 10px;
  height: 30px;
}

.App-header {
  background-color: #bbb;
  padding: 15px;
  color: white;
}

.tableHolder {
  margin-top: 20px;
}

.button {
  width: 200px;
  background-color: #4caf50;
  border: none;
  color: white;
  padding: 10px 15px;
  text-align: center;
  font-size: 15px;
}

.button:hover {
  background-color: #00bb88;
  color: white;
}

.buttonBack {
  margin-top: 20px;
  width: 200px;
  background-color: #4236eb;
  border: none;
  color: white;
  padding: 10px 15px;
  text-align: center;
  font-size: 15px;
}

.buttonBack:hover {
  background-color: #335eec;
  color: white;
}

.cryptoCurrencyTable {
  margin-left: auto;
  margin-right: auto;
  margin-top: 30px;
}

.tableRow {
  background-color: #fff;
  border: none;
  color: #000;
  padding: 10px 15px;
  text-align: center;
  font-size: 14px;
  border-width: 0px;
  cursor: pointer;
}

.tableRow:hover {
  background-color: #eee;
  color: #000;
  border-width: 0px;
  cursor: pointer;
}
