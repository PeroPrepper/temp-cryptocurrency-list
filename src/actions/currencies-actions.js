import axios from "axios";
export const UPDATE_CURRENCIES = "currencies:updateAllCurrencies";
export const CURRENCIES_SET_DEFAULT_FIAT = "currencies:setDefaultFiat";

export function updateAllCurrencies(newCurrenciesData) {
  return {
    type: UPDATE_CURRENCIES,
    payload: {
      currencyList: newCurrenciesData.data,
      fiat: newCurrenciesData.fiat
    }
  };
}

export function showLoadingFetchingCurrencies() {
  return {
    type: UPDATE_CURRENCIES,
    payload: { loading: true }
  };
}

export function showErrorFetchingCurrencies(errorMsg) {
  return {
    type: UPDATE_CURRENCIES,
    payload: { error: errorMsg }
  };
}

export function setDefaultFiat(newFiat) {
  return {
    type: CURRENCIES_SET_DEFAULT_FIAT,
    payload: {
      fiat: newFiat
    }
  };
}

export function fetchAllCurrencies(fiatCurrency) {
  return dispatch => {
    dispatch(showLoadingFetchingCurrencies());

    axios
      .get(
        "https://api.coinmarketcap.com/v1/ticker/?convert=" +
          fiatCurrency +
          "&start=0&limit=100"
      )
      .then(function(response) {
        dispatch(
          updateAllCurrencies({ data: response.data, fiat: fiatCurrency })
        );
      })
      .catch(function(error) {
        dispatch(showErrorFetchingCurrencies(error));
      });
  };
}
