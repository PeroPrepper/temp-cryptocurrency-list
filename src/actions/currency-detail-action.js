import axios from "axios";
export const FETCH_CURRENCY_DETAIL = "detailed-currency:fetchCurrencyDetail";

export function updateCurrencyDetail(newCurrencyData) {
  return {
    type: FETCH_CURRENCY_DETAIL,
    payload: {
      currencyDetailInfo: newCurrencyData.data,
      fiat: newCurrencyData.fiat
    }
  };
}

export function showLoadingFetchingCurrencyDetail() {
  return {
    type: FETCH_CURRENCY_DETAIL,
    payload: { loading: true }
  };
}

export function showErrorFetchingCurrencyDetail(errorMsg) {
  return {
    type: FETCH_CURRENCY_DETAIL,
    payload: { error: errorMsg }
  };
}

export function fetchCurrencyDetail(currencyId, fiatCurrency) {
  return dispatch => {
    dispatch(showLoadingFetchingCurrencyDetail());

    axios
      .get(
        "https://api.coinmarketcap.com/v1/ticker/" +
          currencyId +
          "/?convert=" +
          fiatCurrency
      )
      .then(function(response) {
        dispatch(
          updateCurrencyDetail({ data: response.data[0], fiat: fiatCurrency })
        );
      })
      .catch(function(error) {
        dispatch(showErrorFetchingCurrencyDetail(error));
      });
  };
}

export function setCurrencyDetail(newCurrencyData, pFiat) {
  return {
    type: FETCH_CURRENCY_DETAIL,
    payload: { currencyDetailInfo: newCurrencyData, fiat: pFiat }
  };
}
