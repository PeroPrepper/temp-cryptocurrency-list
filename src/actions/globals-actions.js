export const SET_DETAILED_CURRENCY = "globals:setDetailedCurrency";
export const SET_FIAT_CURRENCY = "globals:setFiatCurrency";

export function setDetailedCurrency(detailedCurrencyId) {
  return {
    type: SET_DETAILED_CURRENCY,
    payload: detailedCurrencyId
  };
}

export function setFiatCurrency(fiatCurrency) {
  return {
    type: SET_FIAT_CURRENCY,
    payload: fiatCurrency
  };
}
