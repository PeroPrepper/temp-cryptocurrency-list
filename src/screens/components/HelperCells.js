import React, { Component } from "react";

class CellChange extends Component {
  render() {
    const { change } = this.props;

    if (Number(change) >= 0) {
      return <b style={{ color: "#009933" }}>{change}%</b>;
    } else {
      return <b style={{ color: "#ff0000" }}>{change}%</b>;
    }
  }
}

export { CellChange };
