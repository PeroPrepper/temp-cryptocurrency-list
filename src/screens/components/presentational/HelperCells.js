import React from "react";

export const CellChange = props => {
  if (Number(props.change) >= 0) {
    return <b style={{ color: "#009933" }}>{props.change}%</b>;
  } else {
    return <b style={{ color: "#ff0000" }}>{props.change}%</b>;
  }
};
