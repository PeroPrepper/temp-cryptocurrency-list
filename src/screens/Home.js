import React from "react";
import { connect } from "react-redux";

import CryptocurrencyTable from "./components/CryptocurrencyTable";
import { fetchAllCurrencies } from "../actions/currencies-actions";

class Home extends React.Component {
  componentDidMount() {
    /* Fetch data from API if the list is empty or if there was a change in the fiat currency selection */
    if (
      this.props.currencies.currencyList.length === 0 ||
      this.props.globals.fiatCurrency !== this.props.currencies.fiat
    ) {
      this.props.onFetchAllCurrencies(this.props.globals.fiatCurrency);
    }
  }

  render() {
    return (
      <div>
        <h1> Cryptocurrency List</h1>
        <button
          id="button"
          onClick={() =>
            this.props.onFetchAllCurrencies(this.props.globals.fiatCurrency)
          }
          className="button"
        >
          Refresh
        </button>
        <div className="tableHolder">
          <CryptocurrencyTable />
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    currencies: state.currencies,
    globals: state.globals
  };
};

const mapActionsToProps = {
  onFetchAllCurrencies: fetchAllCurrencies
};

export default connect(mapStateToProps, mapActionsToProps)(Home);
