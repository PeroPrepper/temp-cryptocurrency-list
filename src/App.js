import React, { Component } from "react";
import "./App.css";

import { Link } from "react-router";

import settingsIcon from "./images/settings.svg";
import homeIcon from "./images/home.svg";
import { getCookie } from "redux-cookie";
import { setFiatCurrency } from "../src/actions/globals-actions";
import { setDefaultFiat } from "../src/actions/currencies-actions";

import { connect } from "react-redux";

class App extends Component {
  /* If cookie exists, extract fiat currency info and save it to store. */
  componentWillMount() {
    var fiatCookie = this.props.onGetCookie("cookieFiatCurrency");
    if (fiatCookie !== undefined) {
      this.props.onSetFiatCurrency(fiatCookie);
      this.props.onSetDefaultFiat(fiatCookie);
    }
  }
  render() {
    return (
      <div className="App">
        <div className="App-header">
          <Link to="/">
            <img className="Menu-icon" src={homeIcon} alt="home" />
          </Link>
          <Link to={"/settings"}>
            <img className="Menu-icon" src={settingsIcon} alt="settings" />
          </Link>
        </div>

        <div>{this.props.children}</div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {};
};

const mapActionsToProps = {
  onGetCookie: getCookie,
  onSetFiatCurrency: setFiatCurrency,
  onSetDefaultFiat: setDefaultFiat
};

export default connect(mapStateToProps, mapActionsToProps)(App);
