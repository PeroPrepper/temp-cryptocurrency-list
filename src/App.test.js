import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import App from "./App";

import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Home } from "./screens/Home";
import { createMockStore } from "redux-test-utils";

import currenciesReducer from "./reducers/currencies-reducer";
import globalsReducer from "./reducers//globals-reducer";
import {
  UPDATE_CURRENCIES,
  CURRENCIES_SET_DEFAULT_FIAT
} from "./actions/currencies-actions";

import { showErrorFetchingCurrencies } from "./actions/currencies-actions";

import { CellChange } from "./screens/components/presentational/HelperCells";
import renderer from "react-test-renderer";

// TEST A COMPONENT
describe("Price Change Component Rendering", () => {
  it("Renders correctly with  positive input value.", () => {
    const tree = renderer.create(<CellChange change={10} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Renders correctly with passed negative input value.", () => {
    const tree = renderer.create(<CellChange change={-15} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// TEST A REDUCER
describe("Currencies Reducer", () => {
  it("Should return the initial state when no action type is provided", () => {
    expect(currenciesReducer("SomeState", {})).toEqual("SomeState");
  });

  it("Should return the initial state when unknown action type is provided", () => {
    const someAction = {
      type: "WrongType",
      payload: { fiat: "USD" }
    };

    expect(currenciesReducer({ test: true }, someAction)).toEqual({
      test: true
    });
  });

  it("Should handle UPDATE_CURRENCIES action and return the payload", () => {
    const updateCurrenciesAction = {
      type: UPDATE_CURRENCIES,
      payload: { fiat: "USD" }
    };
    // it's empty on purpose because it's just starting to fetch posts
    expect(currenciesReducer({ test: true }, updateCurrenciesAction)).toEqual({
      fiat: "USD"
    });
  });

  it("Should handle CURRENCIES_SET_DEFAULT_FIAT action and return previos state + payload", () => {
    const setDefaultFiat = {
      type: CURRENCIES_SET_DEFAULT_FIAT,
      payload: { fiat: "USD" }
    };
    // it's empty on purpose because it's just starting to fetch posts
    expect(currenciesReducer({ test: true }, setDefaultFiat)).toEqual({
      fiat: "USD",
      test: true
    });
  });
});

// TEST AN ACTION
describe("Currencies Error Action", () => {
  it("Should return an error object with correct type and message.", () => {
    expect(showErrorFetchingCurrencies("Error Message")).toEqual({
      type: UPDATE_CURRENCIES,
      payload: { error: "Error Message" }
    });
  });
});
