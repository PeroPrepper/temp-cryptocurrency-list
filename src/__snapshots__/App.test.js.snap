// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price Change Component Rendering Renders correctly with  positive input value. 1`] = `
<b
  style={
    Object {
      "color": "#009933",
    }
  }
>
  10
  %
</b>
`;

exports[`Price Change Component Rendering Renders correctly with passed negative input value. 1`] = `
<b
  style={
    Object {
      "color": "#ff0000",
    }
  }
>
  -15
  %
</b>
`;
